File: examples/dekita/.claude/hooks/multi-issue-guard.py
#!/usr/bin/env python3
"""1つのworktree/PRで複数Issueを同時に対応しようとした場合に警告する。

Why:
    1 worktree = 1 Issue、1 PR = 1 Issueの原則を維持することで、
    変更の追跡性とレビューの容易さを確保する。

What:
    - git worktree addのブランチ名/パスから複数Issue番号を検出
    - gh pr createのbodyから複数Closes/Fixes/Resolvesを検出
    - 複数Issueを検出したらsystemMessageで警告

Remarks:
    - 警告型フック（ブロックしない、正当なケースがあるため）
    - PreToolUse:Bashで発火（git worktree add、gh pr create）
    - issue-auto-assignは単一Issue検出（責務分離）
    - 関連Issue同時修正や親子関係など正当なケースも存在

Changelog:
    - silenvx/dekita#xxx: フック追加
"""

import json
import os
import re

from lib.execution import log_hook_execution
from lib.results import make_approve_result
from lib.session import parse_hook_input
from lib.strings import extract_inline_skip_env, is_skip_env_enabled

SKIP_ENV = "SKIP_MULTI_ISSUE_GUARD"
HOOK_NAME = "multi-issue-guard"


def extract_all_issue_numbers(text: str) -> list[int]:
    """Extract all unique Issue numbers from text.

    Uses the same patterns as issue-auto-assign.py but collects all matches.

    Args:
        text: Text to search for Issue numbers.

    Returns:
        List of unique Issue numbers found (sorted for deterministic output).
    """
    patterns = [
        r"#(\d+)",  # #123
        r"issue[_-](\d+)",  # issue-123, issue_123
        r"/(\d+)[-_]",  # /123-description
        r"[-_](\d+)[-_]",  # feature-123-name
        r"[-_](\d+)$",  # feature-123 (at end)
    ]

    issue_numbers: set[int] = set()
    for pattern in patterns:
        for match in re.finditer(pattern, text, re.IGNORECASE):
            issue_numbers.add(int(match.group(1)))

    return sorted(issue_numbers)


def extract_closing_issue_numbers(body: str) -> list[int]:
    """Extract Issue numbers from closing keywords in PR body.

    Args:
        body: PR body text.

    Returns:
        List of unique Issue numbers from closing keywords.
    """
    pattern = r"(?:closes|fixes|resolves)\s*#?(\d+)"
    matches = re.findall(pattern, body, re.IGNORECASE)
    return sorted(set(int(m) for m in matches))


def parse_worktree_add_command(command: str) -> tuple[str | None, str | None]:
    """Parse git worktree add command and extract branch name and path.

    Copied from issue-auto-assign.py for consistency.

    Returns:
        Tuple of (branch_name, worktree_path).
    """
    if "git worktree add" not in command:
        return None, None

    branch_name = None
    worktree_path = None

    # Look for -b <branch> pattern
    branch_match = re.search(r"-b\s+([^\s]+)", command)
    if branch_match:
        branch_name = branch_match.group(1)

    # Parse command parts to find positional arguments
    parts = command.split()

    try:
        add_idx = parts.index("add")
    except ValueError:
        return branch_name, worktree_path

    # Collect positional arguments (non-option arguments after 'add')
    positional_args = []
    skip_next = False
    for part in parts[add_idx + 1 :]:
        if skip_next:
            skip_next = False
            continue
        if part.startswith("-"):
            if part in ("-b", "--reason"):
                skip_next = True
            continue
        positional_args.append(part)

    if len(positional_args) >= 1:
        worktree_path = positional_args[0]

    if len(positional_args) >= 2 and not branch_name:
        branch_name = positional_args[1]

    return branch_name, worktree_path


def extract_pr_body(command: str) -> str | None:
    """Extract --body argument from gh pr create command.

    Args:
        command: The full command string.

    Returns:
        The body content if found, None otherwise.
    """
    # Match --body "..." or --body '...'
    # Handle escaped quotes within
    patterns = [
        r'--body\s+"((?:[^"\\]|\\.)*)"',  # --body "..."
        r"--body\s+'((?:[^'\\]|\\.)*)'",  # --body '...'
    ]

    for pattern in patterns:
        match = re.search(pattern, command, re.DOTALL)
        if match:
            return match.group(1)

    return None


def check_worktree_command(command: str) -> dict:
    """Check git worktree add command for multiple Issues.

    Args:
        command: The command string.

    Returns:
        Dict with 'warn' (bool) and 'message' (str) keys.
    """
    if "git worktree add" not in command:
        return {"warn": False, "message": ""}

    branch_name, worktree_path = parse_worktree_add_command(command)

    # Collect Issue numbers from both branch and path
    all_issues: set[int] = set()

    if branch_name:
        all_issues.update(extract_all_issue_numbers(branch_name))

    if worktree_path:
        all_issues.update(extract_all_issue_numbers(worktree_path))

    if len(all_issues) > 1:
        issue_list = ", ".join(f"#{i}" for i in sorted(all_issues))
        message = (
            f"⚠️ 複数Issueの同時対応を検出: {issue_list}\n\n"
            f"1 worktree = 1 Issue を推奨します。\n"
            f"本当に複数Issueを同時に対応しますか？\n\n"
            f"意図的な場合は続行してください。"
        )
        return {"warn": True, "message": message, "issues": sorted(all_issues)}

    return {"warn": False, "message": ""}


def check_pr_command(command: str) -> dict:
    """Check gh pr create command for multiple closing keywords.

    Args:
        command: The command string.

    Returns:
        Dict with 'warn' (bool) and 'message' (str) keys.
    """
    if "gh pr create" not in command:
        return {"warn": False, "message": ""}

    body = extract_pr_body(command)
    if not body:
        return {"warn": False, "message": ""}

    issue_numbers = extract_closing_issue_numbers(body)

    if len(issue_numbers) > 1:
        issue_list = ", ".join(f"#{i}" for i in issue_numbers)
        message = (
            f"⚠️ 複数Issueを同時にクローズしようとしています: {issue_list}\n\n"
            f"1 PR = 1 Issue を推奨します。\n"
            f"複数Issueを同時にクローズするのは意図的ですか？\n\n"
            f"意図的な場合は続行してください。"
        )
        return {"warn": True, "message": message, "issues": issue_numbers}

    return {"warn": False, "message": ""}


def main() -> None:
    """Entry point for the hook."""
    try:
        input_data = parse_hook_input()
    except (json.JSONDecodeError, Exception):
        # Fail open: allow on parse errors
        return

    tool_name = input_data.get("tool_name", "")
    if tool_name != "Bash":
        return  # Not a Bash command

    tool_input = input_data.get("tool_input", {})
    command = tool_input.get("command", "")

    if not command:
        return  # Empty command

    # Check SKIP environment variable (exported)
    if is_skip_env_enabled(os.environ.get(SKIP_ENV)):
        log_hook_execution(HOOK_NAME, "skip", f"{SKIP_ENV}=1: チェックをスキップ")
        result = make_approve_result(HOOK_NAME)
        print(json.dumps(result))
        return

    # Check inline SKIP environment variable
    inline_value = extract_inline_skip_env(command, SKIP_ENV)
    if is_skip_env_enabled(inline_value):
        log_hook_execution(HOOK_NAME, "skip", f"{SKIP_ENV}=1: チェックをスキップ（インライン）")
        result = make_approve_result(HOOK_NAME)
        print(json.dumps(result))
        return

    # Check worktree add command
    worktree_result = check_worktree_command(command)
    if worktree_result["warn"]:
        log_hook_execution(
            HOOK_NAME,
            "warn",
            f"複数Issue検出（worktree）: {worktree_result.get('issues', [])}",
        )
        result = {"decision": "approve", "systemMessage": worktree_result["message"]}
        print(json.dumps(result))
        return

    # Check PR create command
    pr_result = check_pr_command(command)
    if pr_result["warn"]:
        log_hook_execution(
            HOOK_NAME,
            "warn",
            f"複数Issue検出（PR）: {pr_result.get('issues', [])}",
        )
        result = {"decision": "approve", "systemMessage": pr_result["message"]}
        print(json.dumps(result))
        return

    # No warning needed - output nothing (per hooks-reference pattern)


if __name__ == "__main__":
    main()

File: examples/dekita/.claude/hooks/handlers/multi_issue_guard.ts
#!/usr/bin/env bun
/**
 * 1つのworktree/PRで複数Issueを同時に対応しようとした場合に警告する。
 *
 * Why:
 *   1 worktree = 1 Issue、1 PR = 1 Issueの原則を維持することで、
 *   変更の追跡性とレビューの容易さを確保する。
 *
 * What:
 *   - git worktree addのブランチ名/パスから複数Issue番号を検出
 *   - gh pr createのbodyから複数Closes/Fixes/Resolvesを検出
 *   - 複数Issueを検出したらsystemMessageで警告
 *
 * Remarks:
 *   - 警告型フック（ブロックしない、正当なケースがあるため）
 *   - PreToolUse:Bashで発火（git worktree add、gh pr create）
 *   - issue-auto-assignは単一Issue検出（責務分離）
 *   - 関連Issue同時修正や親子関係など正当なケースも存在
 *
 * Changelog:
 *   - silenvx/dekita#2932: TypeScriptに移行
 */

import { extractPrBody } from "../lib/command";
import { logHookExecution } from "../lib/logging";
import { makeApproveResult } from "../lib/results";
import { createHookContext, parseHookInput } from "../lib/session";
import { extractInlineSkipEnv, isSkipEnvEnabled } from "../lib/strings";

const SKIP_ENV = "SKIP_MULTI_ISSUE_GUARD";
const HOOK_NAME = "multi-issue-guard";

interface CheckResult {
  warn: boolean;
  message: string;
  issues?: number[];
}

/**
 * Extract all unique Issue numbers from text.
 *
 * Uses the same patterns as issue-auto-assign but collects all matches.
 */
export function extractAllIssueNumbers(text: string): number[] {
  const patterns = [
    /#(\d+)/g, // #123
    /issue[_-](\d+)/gi, // issue-123, issue_123
    /\/(\d+)[-_]/g, // /123-description
    /[-_](\d+)[-_]/g, // feature-123-name
    /[-_](\d+)$/g, // feature-123 (at end)
  ];

  const issueNumbers = new Set<number>();
  for (const pattern of patterns) {
    // Use matchAll to avoid assignment in expression
    for (const match of text.matchAll(pattern)) {
      issueNumbers.add(Number.parseInt(match[1], 10));
    }
  }

  return [...issueNumbers].sort((a, b) => a - b);
}

/**
 * Extract Issue numbers from closing keywords in PR body.
 */
export function extractClosingIssueNumbers(body: string): number[] {
  const pattern = /(?:closes|fixes|resolves)\s*#?(\d+)/gi;
  const matches: number[] = [];
  // Use matchAll to avoid assignment in expression
  for (const match of body.matchAll(pattern)) {
    matches.push(Number.parseInt(match[1], 10));
  }
  return [...new Set(matches)].sort((a, b) => a - b);
}

/**
 * git worktree addのうち、引数を取る既知のオプション。
 *
 * これらのオプションは「次のトークン」を引数として消費するため、
 * パス/ブランチ名として誤認識しないよう、パース時にスキップする。
 *
 * 制限事項:
 *   - `--track=direct` のような `--opt=value` 形式は、1トークンで完結するため
 *     ここでは特別扱いしていない（`--track` のみを次トークン消費対象として扱う）。
 *   - 将来 git に新しい引数付きオプションが追加された場合は、
 *     必要に応じてこのセットを更新すること。
 */
const OPTIONS_WITH_ARG = new Set(["-b", "-B", "--reason", "--orphan"]);

/**
 * Tokenize a shell command, respecting quoted strings and escape sequences.
 *
 * Shell escape rules:
 *   - Single quotes: No escapes processed (all characters literal)
 *   - Double quotes: Only $ ` " \ \n are escapable
 *   - Unquoted: All backslash escapes processed
 *   - Line continuation: \+newline removed (outside double quotes)
 *   - Trailing backslash: Preserved as literal (shell would error/wait)
 */
function tokenizeCommand(command: string): string[] {
  const tokens: string[] = [];
  let current = "";
  let inSingleQuote = false;
  let inDoubleQuote = false;

  for (let i = 0; i < command.length; i++) {
    const char = command[i];

    // Handle backslash escapes (outside single quotes)
    // ダブルクォート内では $ ` " \ \n のみがエスケープ対象
    // その他の文字（例: \P）はバックスラッシュがリテラルとして保持される
    if (char === "\\" && !inSingleQuote) {
      if (i + 1 < command.length) {
        const nextChar = command[i + 1];
        // Line continuation (backslash + newline) - remove both characters
        // シェルでは行継続として処理される（クォート外でも内でも）
        if (nextChar === "\n" && !inDoubleQuote) {
          i++;
          continue;
        }
        // Inside double quotes, only specific characters are escaped
        if (inDoubleQuote && !["$", "`", '"', "\\", "\n"].includes(nextChar)) {
          current += "\\";
          continue;
        }
        i++;
        current += command[i];
      } else {
        // Trailing backslash: preserve as literal
        // (shell would error or wait for more input)
        current += "\\";
      }
      continue;
    }

    if (char === "'" && !inDoubleQuote) {
      inSingleQuote = !inSingleQuote;
      continue;
    }

    if (char === '"' && !inSingleQuote) {
      inDoubleQuote = !inDoubleQuote;
      continue;
    }

    // Split on any whitespace (space, tab, newline)
    if (
      (char === " " || char === "\t" || char === "\n" || char === "\r") &&
      !inSingleQuote &&
      !inDoubleQuote
    ) {
      if (current) {
        tokens.push(current);
        current = "";
      }
      continue;
    }

    current += char;
  }

  if (current) {
    tokens.push(current);
  }

  return tokens;
}

/**
 * Parse git worktree add command and extract branch name and path.
 */
export function parseWorktreeAddCommand(command: string): {
  branchName: string | null;
  worktreePath: string | null;
} {
  if (!command.includes("git worktree add")) {
    return { branchName: null, worktreePath: null };
  }

  let branchName: string | null = null;
  let worktreePath: string | null = null;

  // Tokenize command respecting quotes
  const parts = tokenizeCommand(command);
  const addIdx = parts.indexOf("add");
  if (addIdx === -1) {
    return { branchName, worktreePath };
  }

  // Collect positional arguments (non-option arguments after 'add')
  const positionalArgs: string[] = [];
  let expectingBranch = false;
  let skipNext = false;
  for (const part of parts.slice(addIdx + 1)) {
    if (skipNext) {
      if (expectingBranch) {
        branchName = part;
        expectingBranch = false;
      }
      skipNext = false;
      continue;
    }
    if (part.startsWith("-")) {
      // Check if this option takes an argument
      if (OPTIONS_WITH_ARG.has(part)) {
        skipNext = true;
        if (part === "-b" || part === "-B" || part === "--orphan") {
          expectingBranch = true;
        }
      }
      continue;
    }
    positionalArgs.push(part);
  }

  if (positionalArgs.length >= 1) {
    worktreePath = positionalArgs[0];
  }

  if (positionalArgs.length >= 2 && !branchName) {
    branchName = positionalArgs[1];
  }

  return { branchName, worktreePath };
}

// extractPrBody is imported from ../lib/command

/**
 * Check git worktree add command for multiple Issues.
 */
export function checkWorktreeCommand(command: string): CheckResult {
  if (!command.includes("git worktree add")) {
    return { warn: false, message: "" };
  }

  const { branchName, worktreePath } = parseWorktreeAddCommand(command);

  // Collect Issue numbers from both branch and path
  const allIssues = new Set<number>();

  if (branchName) {
    for (const issue of extractAllIssueNumbers(branchName)) {
      allIssues.add(issue);
    }
  }

  if (worktreePath) {
    for (const issue of extractAllIssueNumbers(worktreePath)) {
      allIssues.add(issue);
    }
  }

  if (allIssues.size > 1) {
    const sortedIssues = [...allIssues].sort((a, b) => a - b);
    const issueList = sortedIssues.map((i) => `#${i}`).join(", ");
    const message = `⚠️ 複数Issueの同時対応を検出: ${issueList}

1 worktree = 1 Issue を推奨します。
本当に複数Issueを同時に対応しますか？

意図的な場合は続行してください。`;
    return { warn: true, message, issues: sortedIssues };
  }

  return { warn: false, message: "" };
}

/**
 * Check gh pr create command for multiple closing keywords.
 */
export function checkPrCommand(command: string): CheckResult {
  if (!command.includes("gh pr create")) {
    return { warn: false, message: "" };
  }

  const body = extractPrBody(command);
  if (!body) {
    return { warn: false, message: "" };
  }

  const issueNumbers = extractClosingIssueNumbers(body);

  if (issueNumbers.length > 1) {
    const issueList = issueNumbers.map((i) => `#${i}`).join(", ");
    const message = `⚠️ 複数Issueを同時にクローズしようとしています: ${issueList}

1 PR = 1 Issue を推奨します。
複数Issueを同時にクローズするのは意図的ですか？

意図的な場合は続行してください。`;
    return { warn: true, message, issues: issueNumbers };
  }

  return { warn: false, message: "" };
}

async function main(): Promise<void> {
  let inputData: Awaited<ReturnType<typeof parseHookInput>>;
  let sessionId: string | undefined;
  try {
    inputData = await parseHookInput();
    const ctx = createHookContext(inputData);
    sessionId = ctx.sessionId;
  } catch {
    // Fail open: allow on parse errors
    return;
  }

  const toolName = inputData.tool_name ?? "";
  if (toolName !== "Bash") {
    return; // Not a Bash command
  }

  const toolInput = inputData.tool_input ?? {};
  const command = (toolInput as { command?: string }).command ?? "";

  if (!command) {
    return; // Empty command
  }

  // Check SKIP environment variable (exported)
  if (isSkipEnvEnabled(process.env[SKIP_ENV])) {
    await logHookExecution(HOOK_NAME, "skip", `${SKIP_ENV}=1: チェックをスキップ`, undefined, {
      sessionId,
    });
    const result = makeApproveResult(HOOK_NAME);
    console.log(JSON.stringify(result));
    return;
  }

  // Check inline SKIP environment variable
  const inlineValue = extractInlineSkipEnv(command, SKIP_ENV);
  if (isSkipEnvEnabled(inlineValue)) {
    await logHookExecution(
      HOOK_NAME,
      "skip",
      `${SKIP_ENV}=1: チェックをスキップ（インライン）`,
      undefined,
      { sessionId },
    );
    const result = makeApproveResult(HOOK_NAME);
    console.log(JSON.stringify(result));
    return;
  }

  // Check worktree add command
  const worktreeResult = checkWorktreeCommand(command);
  if (worktreeResult.warn) {
    await logHookExecution(
      HOOK_NAME,
      "warn",
      `複数Issue検出（worktree）: ${JSON.stringify(worktreeResult.issues)}`,
      undefined,
      { sessionId },
    );
    const result = { systemMessage: worktreeResult.message };
    console.log(JSON.stringify(result));
    return;
  }

  // Check PR create command
  const prResult = checkPrCommand(command);
  if (prResult.warn) {
    await logHookExecution(
      HOOK_NAME,
      "warn",
      `複数Issue検出（PR）: ${JSON.stringify(prResult.issues)}`,
      undefined,
      { sessionId },
    );
    const result = { systemMessage: prResult.message };
    console.log(JSON.stringify(result));
    return;
  }

  // No warning needed - output nothing (per hooks-reference pattern)
}

if (import.meta.main) {
  main().catch((error) => {
    console.error(`[${HOOK_NAME}] Error:`, error);
    process.exit(0); // Fail open
  });
}
